### PR TITLE
Contribution guidelines

### PR DESCRIPTION
**Pull request by samuelson**
This is part of the project to use a release branch.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+maintain passing states in our testing pipelines.
+
+The original contributor will be notified of the revert in the Jira ticket
+associated with the change. A reference to the test(s) and operating system(s)
+that failed as a result of the code change will also be added to the Jira
+ticket. This test(s) should be used to check future submissions of the code to
+ensure the issue has been resolved.
+
+## Release Policy
+Stable features and bug fixes will be moved to the release branch on a weekly
```
Comment (marrero984) [inline CONTRIBUTING.md]: Is this movement already automated?  I wouldn't want to commit to something weekly unless the automation is already deployed and working reliably

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+````
+
+## Submitting Changes
+
+* Sign the [Contributor License Agreement](http://links.puppet.com/cla).
+* Push your changes to a topic branch in your fork of the repository.
+* Submit a pull request to the repository in the puppetlabs organization.
+* Update your Jira ticket to mark that you have submitted code and are ready for it to be reviewed (Status: Ready for Merge).
+  * Include a link to the pull request in the ticket.
+* The education team looks at Pull Requests on a regular basis.
```
Comment (greglarkin) [inline CONTRIBUTING.md]: Suggest: Education team 
Suggest: Lowercase "Pull Requests" for consistency

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+discretion of the committer and maintainers. This action would be taken to help 
+maintain passing states in our testing pipelines.
+
+The original contributor will be notified of the revert in the Jira ticket
+associated with the change. A reference to the test(s) and operating system(s)
+that failed as a result of the code change will also be added to the Jira
+ticket. This test(s) should be used to check future submissions of the code to
+ensure the issue has been resolved.
+
+## Release Policy
```
Comment (greglarkin) [inline CONTRIBUTING.md]: Dumb question - the "release" branch was created for the Edu repositories. Is there any need for a separate "integration" branches where features are moved and testing is driven from before final push to the "release" branch? Maybe it's not necessary because releases aren't performed very often, but just wondered about it.

Comment (samuelson) [inline CONTRIBUTING.md]: @greglarkin I don't think we need to. Really that's what master should be. My thought is that we can keep the release branch in an always releasable state, but still only publish releases when we choose to.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+* Changes resulting in test pipeline failures will be reverted if they cannot
+  be resolved within one business day.
+
+# Additional Resources
+
+* [Puppet community guidelines](https://docs.puppet.com/community/community_guidelines.html)
+* [Bug tracker (Jira)](https://tickets.puppetlabs.com)
+* [Contributor License Agreement](http://links.puppet.com/cla)
+* [General GitHub documentation](https://help.github.com/)
+* [GitHub pull request documentation](https://help.github.com/send-pull-requests/)
```
Comment (samuelson) [inline CONTRIBUTING.md]: ugh, that's copy/paste from the puppet one, I'll fix it.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+* Make sure you have a [GitHub account](https://github.com/signup/free)
+* Submit a ticket for your issue, assuming one does not already exist.
+  * Clearly describe the issue including steps to reproduce when it is a bug.
+  * Make sure you fill in the earliest version that you know has the issue.
+* Fork the repository on GitHub
+
+## Making Changes
+
+Releases are made from the `release` branch, which should always be in a
+releasable state. `master` is the active development branch and changes
```
Comment (samuelson) [inline CONTRIBUTING.md]: I think it's a pretty standard way of doing gitflow to just have one dev branch and one release branch, but maybe it should say that the integration testing happens in master and that acceptance testing happens in release?

Comment (binford2k) [inline CONTRIBUTING.md]: Oh, I see what you mean. Yes, integration tests may still be pending, but there's not really a way to atomically update and test all the projects simultaneously. Instead, the release branch is always tested and releasable *on its own*, and the tagged releases are the ones that have been integration tested and actually released.

The release branch is not intended to be used by external parties.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+already been addressed in the `master` branch.
+
+* Create a topic branch from where you want to base your work.
+  * This is usually the master branch.
+  * Only target release branches if you are certain your fix must be on that
+    branch.
+  * To quickly create a topic branch based on master; `git checkout -b
+    fix/master/my_contribution master`. Please avoid working directly on the
+    `master` branch.
+* Make commits of logical units.
```
Comment (binford2k) [inline CONTRIBUTING.md]: Make atomic commits of logical units. Do not combine unrelated fixes.

File: CONTRIBUTING.md (hunk 5 above)
Comment (samuelson) [inline CONTRIBUTING.md]: actually on rereading it I think that's clear enough

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+* Check for unnecessary whitespace with `git diff --check` before committing.
+* Make sure your commit messages are in the proper format.
```
Comment (binford2k) [inline CONTRIBUTING.md]: * Don't update version numbers or changelogs. Instead, write a good commit subject and description. We will use the commit subject to update the changelog as needed.

Do you like that @samuelson ?

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    based on a description rather than an example.  This patch fixes the
+    problem by making the example concrete and imperative.
+
+    The first line is a real life imperative statement with a ticket number
+    from our issue tracker.  The body describes the behavior without the patch,
+    why this is a problem, and how the patch fixes the problem when applied.
+````
+
+* Make sure you have added the necessary tests for your changes.
+* Run _all_ the tests to assure nothing else was accidentally broken.
```
Comment (binford2k) [inline CONTRIBUTING.md]: show a quick example of how to run the tests, and/or link to a document on how to set up and use a testing environment for our projects.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+## Revert Policy
+By running tests in advance and by engaging with peer review for prospective
+changes, your contributions have a high probability of becoming long lived
+parts of the the project. After being merged, the code will run through a
+series of testing pipelines on a large number of operating system
+environments. These pipelines can reveal incompatibilities that are difficult
+to detect in advance.
+
+If the code change results in a test failure, we will make our best effort to
+correct the error. If a fix cannot be determined and committed within 24 hours
```
Comment (binford2k) [inline CONTRIBUTING.md]: or as needed for expediency

File: CONTRIBUTING.md (hunk 5 above)
Comment (binford2k) [inline CONTRIBUTING.md]: agreed. I'm a little hesitant to be *too* prescriptive right off the bat. I'm ok with leaving some things a little vague so we have a little flexibility to find out what works.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+### Documentation
+
+For changes of a trivial nature to comments and documentation, it is not
+always necessary to create a new ticket in Jira. In this case, it is
+appropriate to start the first line of a commit with '(doc)' instead of
+a ticket number.
+
+````
+    (doc) Add documentation commit example to CONTRIBUTING
```
Comment (binford2k) [inline CONTRIBUTING.md]: we need to decide !!and actually implement!! the Jira tagging method we want to use. I don't care what it is, but it should be consistent.

File: CONTRIBUTING.md (hunk 8 above)
Comment (samuelson) [inline CONTRIBUTING.md]: Do we have a good resource for this? If not, I'll try to look in other repos to see if I can find something.

File: CONTRIBUTING.md (hunk 10 above)
Comment (samuelson) [inline CONTRIBUTING.md]: lol, I think about half of your comments are from things I copied from the puppet repo. I didn't review it too closely since I figured it would be mostly used outside our team.

File: CONTRIBUTING.md (hunk 8 above)
Comment (binford2k) [inline CONTRIBUTING.md]: Not as far as I know. For now, I guess that just adding a bullet like this is good enough:

    * `rake spec*

Comment (samuelson) [inline CONTRIBUTING.md]: Ok, I'm going to do that and propose we have another doc about how to write and run tests.
